cli/src/commands/package/upload.rs

```diff
@@ -0,0 +1,515 @@
+use chrono::Utc;
+use colored::Colorize;
+use itertools::Itertools;
+use std::{
+    collections::HashMap,
+    fmt::Display,
+    path::PathBuf,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        mpsc::channel,
+        Arc,
+    },
+    thread::sleep,
+    time::Duration,
+};
+
+use anyhow::{anyhow, Context, Result};
+use reinfer_client::{
+    resources::dataset::IxpDatasetNew, Client, CommentUid, Dataset, DatasetFullName, DatasetName,
+    LabelDef, NewAnnotatedComment, NewLabelDef, Source, SourceId, SourceKind,
+    DEFAULT_LABEL_GROUP_NAME,
+};
+use scoped_threadpool::Pool;
+use structopt::StructOpt;
+
+use crate::{
+    commands::package::{AttachmentKey, CommentBatchKey, Package},
+    progress::Progress,
+};
+
+#[derive(Debug, StructOpt)]
+pub struct UploadPackageArgs {
+    /// Path of the package to upload
+    #[structopt(parse(from_os_str))]
+    file: PathBuf,
+
+    #[structopt(long = "resume-on-error")]
+    /// Whether to attempt to resume processing on error
+    resume_on_error: bool,
+
+    #[structopt(long = "max-attachment-memory-mb", default_value = "256")]
+    /// Maximum number of mb to use to store attachments in memory
+    max_attachment_memory_mb: u64,
+
+    #[structopt(long = "dataset-creation-timeout", default_value = "30")]
+    /// Maximum number of seconds to wait for dataset to be created
+    dataset_creation_timeout: i64,
+}
+
+fn wait_for_dataset_to_exist(dataset: &Dataset, client: &Client, timeout_s: i64) -> Result<()> {
+    let start_time = Utc::now();
+
+    while (start_time - Utc::now()).num_seconds() <= timeout_s {
```

Comment:
You should be using the `Instant` type for this kind of call: https://doc.rust-lang.org/std/time/struct.Instant.html

Generally for anything that's performance or timing related that doesn't relate to human times, these are more reliable and faster than anything to do with human time (system clocks can go backwards etc.). Also simpler!

```rust
    let start = Instant::now();
    while start.elapsed().as_secs() <= timeout_s {
```